### PR TITLE
Add buffer for allocation constraints and a switch to disable index b…

### PR DESCRIPTION
**Pull request by gbbafna**
…alance constraint

### Description
Add buffer for allocation constraints and a switch to disable index balance constraints

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 1]
```diff
@@ -135,7 +136,7 @@ public class BalancedShardsAllocator implements ShardsAllocator {
     );
 
     /**
-     * This setting governs whether primary shards balance is desired during allocation. This is used by {@link ConstraintTypes#isPerIndexPrimaryShardsPerNodeBreached()}
+     * This setting governs whether primary shards balance is desired during allocation. This is used by
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:139]: will fix this .

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 2]
```diff
@@ -147,13 +148,44 @@ public class BalancedShardsAllocator implements ShardsAllocator {
...
 
+    public static final Setting<Float> PREFER_PRIMARY_SHARD_BALANCE_BUFFER = Setting.floatSetting(
+        "cluster.routing.allocation.balance.primary.shard.buffer",
+        0.0f,
+        0.0f,
+        Property.Dynamic,
+        Property.NodeScope
+    );
+
+    public static final Setting<Float> PREFER_PRIMARY_INDEX_BALANCE_BUFFER = Setting.floatSetting(
```
Comment (Arpit-Bandejiya) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:173]: Can we keep the setting name coherent with the name of constraints:
```
    public AllocationConstraints(AllocationParameter allocationParameter) {
        this.constraints = new HashMap<>();
        this.constraints.put(
            INDEX_SHARD_PER_NODE_BREACH_CONSTRAINT_ID,
            new Constraint(isIndexShardsPerNodeBreached(allocationParameter.getShardBalanceBuffer()))
        );
        this.constraints.put(
            INDEX_PRIMARY_SHARD_BALANCE_CONSTRAINT_ID,
            new Constraint(isPerIndexPrimaryShardsPerNodeBreached(allocationParameter.getPreferPrimaryBalanceIndexBuffer()))
        );
        this.constraints.put(
            CLUSTER_PRIMARY_SHARD_BALANCE_CONSTRAINT_ID,
            new Constraint(isPrimaryShardsPerNodeBreached(allocationParameter.getPreferPrimaryBalanceShardBuffer()))
        );
    }
```

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 3]
```diff
@@ -410,20 +489,34 @@ static class WeightFunction {
...
             float sum = indexBalance + shardBalance;
-            if (sum <= 0.0f) {
+            if (sum > 0.0f) {
+                theta0 = shardBalance / sum;
+                theta1 = indexBalance / sum;
+            } else if (sum == 0.0f) {
+                theta0 = 0;
+                theta1 = 0;
+            } else {
                 throw new IllegalArgumentException("Balance factors must sum to a value > 0 but was: " + sum);
```
Comment (Arpit-Bandejiya) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:508]: nit: `>=`

File: server/src/main/java/org/opensearch/cluster/routing/allocation/Constraint.java [hunk 4]
```diff
@@ -40,6 +40,10 @@ public void setEnable(boolean enable) {
         this.enable = enable;
     }
 
+    public boolean isEnable() {
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/Constraint.java:43]: Can this be package private if only used in unit tests?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/RebalanceConstraints.java [hunk 5]
```diff
@@ -31,7 +31,7 @@ public class RebalanceConstraints {
 
     public RebalanceConstraints(RebalanceParameter rebalanceParameter) {
         this.constraints = new HashMap<>();
-        this.constraints.put(INDEX_PRIMARY_SHARD_BALANCE_CONSTRAINT_ID, new Constraint(isPerIndexPrimaryShardsPerNodeBreached()));
+        this.constraints.put(INDEX_PRIMARY_SHARD_BALANCE_CONSTRAINT_ID, new Constraint(isPerIndexPrimaryShardsPerNodeBreached(0.0f)));
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/RebalanceConstraints.java:34]: Don't we need to pass the value of setting here?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 6]
```diff
@@ -147,13 +147,44 @@ public class BalancedShardsAllocator implements ShardsAllocator {
         Property.NodeScope
     );
 
+    public static final Setting<Boolean> SHARDS_PER_INDEX_BALANCE = Setting.boolSetting(
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:150]: Where is this setting used?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/Constraint.java (hunk 4 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/Constraint.java:43]: done.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java (hunk 6 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:150]: This is used to enable/disable `INDEX_SHARD_PER_NODE_BREACH_CONSTRAINT_ID` constraint . 

```
        this.weightFunction.updateAllocationConstraint(INDEX_SHARD_PER_NODE_BREACH_CONSTRAINT_ID, shardsPerIndexBalance);

```

File: server/src/main/java/org/opensearch/cluster/routing/allocation/RebalanceConstraints.java (hunk 5 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/RebalanceConstraints.java:34]: We are only allowing the buffer in allocation , but not in rebalance , as we want that to happen perfectly on an index level .

File: server/src/main/java/org/opensearch/cluster/routing/allocation/ConstraintTypes.java [hunk 7]
```diff
@@ -66,11 +66,14 @@ public static Predicate<Constraint.ConstraintParams> isIndexShardsPerNodeBreache
      * {@link ConstraintTypes#CONSTRAINT_WEIGHT} is assigned to node resulting in lesser chances of node being selected
      * as allocation or rebalancing target
      */
```
[the remark this answers is not in the document]
Comment (gbbafna) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/ConstraintTypes.java:68]: Done.

**Pull request closed without merging**